File: web/src/api/orderService.ts
import api from "./api";

interface OrderItem {
  name: string;
  price: number;
  quantity: number
}

interface AddOrderItemRequest {
  menu_item_id: number;
  quantity: number;
}

class OrderService {

  async getAllOrders(token: string) {
    const response = await api.get("/order/getAllOrders", {
      headers: { Authorization: `Bearer ${token}` },
    });
    return response.data;
  }

  async getOrderById(token: string, id: number) {
    const response = await api.get(`/order/getOrderById/${id}`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    return response.data;
  }

  async getAllTables(token: string) {
    const response = await api.get("/table/getAllTables", {
      headers: {
        "Authorization": `Bearer ${token}`
      },
    })
    console.log(JSON.stringify(response.data, null, 2))
    return response.data;
  }

  async closeOrder(token: string, id: number) {
    const response = await api.patch(`/order/updateOrderStatus/${id}`, {}, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
    return response.data;
  }
  
  async addOrderItem(token: string, id: number, item: AddOrderItemRequest) {
    const response = await api.post(`/order/addOrderItem/${id}`,
      item,
      {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`
        },
      });
    return response.data;
  }

  async createOrder(token: string, table_id: number, items: OrderItem[] = []) {
    const response = await api.post(`/order/createOrder`,
      {
        table_id,
        items,
      },
      {
        headers: {
          'Authorization': `Bearer ${token}`
        }
      },
    ) 

    return response.data;
  }

  async discardOrder(token: string, id: number) {
    const response = await api.patch(`/order/deleteOrder/${id}`,
      {},
      {
        headers: {
          'Authorization': `Bearer ${token}`
        }
      }
    )

    return response.data
  }

}

export default new OrderService();

File: server/services/orderService.go
package services

import (
	"github.com/MatheusABA/restaurant-project/server/controller/order/dto"
	"github.com/MatheusABA/restaurant-project/server/model"
	"github.com/MatheusABA/restaurant-project/server/repositories"
)

func CreateOrder(req dto.CreateOrderRequest) (*model.Order, error) {
	orderModel := model.Order{
		UserID:  req.UserID,
		TableID: req.TableID,
		Status:  "open",
	}

	for _, item := range req.Items {
		// Busca o MenuItem pelo ID
		menuItem, err := repositories.GetMenuItemByID(item.MenuItemID)
		if err != nil {
			return nil, err
		}
		orderModel.Items = append(orderModel.Items, model.OrderItem{
			MenuItemID: menuItem.ID,
			Name:       menuItem.Name,
			Price:      menuItem.Price,
			Quantity:   item.Quantity,
		})
	}
	return repositories.CreateOrder(&orderModel)
}

func GetAllOrders() ([]model.Order, error) {
	return repositories.GetAllOrders()
}

func GetOrderById(id uint) (*model.Order, error) {
	return repositories.GetOrderById(id)
}

func UpdateOrderStatus(req dto.CloseOrderRequest) error {
	return repositories.UpdateOrderStatus(req.ID)
}

func AddOrderItem(req dto.AddOrderItemRequest) error {
	menuItem, err := repositories.GetMenuItemByID(req.MenuItemID)
	if err != nil {
		return err
	}

	orderItem := model.OrderItem{
		OrderID:    req.ID,
		MenuItemID: menuItem.ID,
		Name:       menuItem.Name,
		Price:      menuItem.Price,
		Quantity:   req.Quantity,
	}
	return repositories.AddOrderItem(orderItem)
}

func DeleteOrder(req dto.DeleteOrderRequest) error {
	return repositories.DeleteOrder(req.ID)
}
